File: lib/day_01.ex
defmodule AdventOfCode2018.Day01 do
  @doc """
  Parse a list of numbers (integers) and add them together
  Returns the sum of all additions

  ## Examples
      iex> AdventOfCode2018.Day01.part1("+1, +1, +1")
      3

      iex> AdventOfCode2018.Day01.part1("+1, +1, -2")
      0

      iex> AdventOfCode2018.Day01.part1("-1, -2, -3")
      -6
  """
  def part1(input) do
    operations = String.split(input, ",", trim: true)

    Enum.reduce(operations, 0, fn(operation, acc) -> acc + (operation |> String.trim |> String.to_integer) end)
  end

  def part2(args) do
  end
end

File: src/day-01.ts
export class DayOne {

    public partOne(input: string): number {
        return this.transformString(input)
            .reduce((accumulator, value) => {
                return accumulator + value
            });
    }

    public partTwo(input: string): number {
        const operations = this.transformString(input);

        const sums = Array.from([0]);

        let lastSum = 0;

        while (true) {
            for (const x of operations) {
                lastSum = sums[sums.length - 1] + x

                if (sums.includes(lastSum)) {
                    return lastSum;
                }

                sums.push(lastSum);
            }
        }
    }

    private transformString(input: string) : Array<number> {
        return input.replace(/\n/g, ",")
            .split(",")
            .map(value => value.trim())
            .map(value => parseInt(value, 10))
    }
}